sortinghat/core/jobs.py

```diff
@@ -443,23 +454,34 @@ def unify(ctx, criteria, source_uuids=None, target_uuids=None, exclude=True, str
     def _group_recommendations(recs):
         """Calculate unique sets of identities from matching recommendations.
 
-        For instance, given a list of matching groups like
-        A = {A, B}; B = {B,A,C}, C = {C,} and D = {D,} the output
-        for keys A, B and C will be the group {A, B, C}. As D has no matches,
-        it won't be included in any group and it won't be returned.
+        For instance, given a dictionary of matching groups like
+        {A: [B], B: [A,C], D: [E]} the output will be the groups
+        [{A, B, C}, {D, E}].
 
         :param recs: recommendations of matching identities
 
         :returns: a list including unique groups of matches
         """
-        groups = []
-        for group_key in recs:
-            g_uuids = pandas.Series(recs[group_key])
-            g_uuids = g_uuids.append(pandas.Series([group_key]))
-            g_uuids = list(g_uuids.sort_values().unique())
-            if (len(g_uuids) > 1) and (g_uuids not in groups):
-                groups.append(g_uuids)
-        return groups
+        visited = set()
+
+        def dfs(node, current_set):
+            # Depth First Search algorithm
+            # Keep visiting the keys until the end
+            if node not in visited:
+                visited.add(node)
+                current_set.add(node)
+                for neighbor in recs.get(node, []):
+                    dfs(neighbor, current_set)
+
+        result = []
+        for key in recs:
+            if key not in visited:
+                current_set = set()
+                dfs(key, current_set)
+                if len(current_set) > 1:
+                    result.append(current_set)
+
```

Comment:
I think this is a bit complicated. I don't like that recursive function and I don't like it's using global variables. A better approach for DFS is to think that you will need to empty a stack of nodes.
Every time you discover a new node, you add it to the stack, on the next iteration you get the first element from the stack. You do that until the stack is empty. When it's empty, you have a path and you start processing a new path using a new empty stack. Because there can be cycles, you should still use the visited set to avoid doing the work twice.

Reply:
I updated the function, let me know if the new algorithm is clearer